libs/auth/src/artifact_authenticator.ts

```diff
@@ -0,0 +1,202 @@
+    const signatureFile = this.serializeSignatureFileContents({
+      signature: messageSignature,
+      signingMachineCert,
+    });
+    await fs.writeFile(
+      this.constructSignatureFilePath(artifact),
+      signatureFile
+    );
+  }
+
+  /**
+   * Verifies the authenticity of the provided artifact using its signature file, which is expected
+   * to be found at the same file path as the artifact, but with a .vxsig extension, e.g.
+   * /path/to/artifact.txt.vxsig.
+   */
+  async authenticateArtifactUsingSignatureFile(
+    artifact: Artifact
+  ): Promise<Result<void, Error>> {
+    try {
+      await this.authenticateArtifactUsingSignatureFileHelper(artifact);
+    } catch {
+      // TODO: Log raw error
+      return err(
+        new Error(`Error authenticating ${artifact.path} using signature file`)
+      );
+    }
+    return ok();
+  }
+
+  private async authenticateArtifactUsingSignatureFileHelper(
+    artifact: Artifact
+  ): Promise<void> {
+    const message = await this.constructMessage(artifact);
+
+    const signatureFile = await fs.readFile(
+      this.constructSignatureFilePath(artifact)
+    );
+    const { signature: messageSignature, signingMachineCert } =
+      this.deserializeSignatureFileContents(signatureFile);
+
+    // Verify type and authenticity of signing machine cert
+    const certDetails = await parseCert(signingMachineCert);
+    switch (artifact.type) {
+      case 'cvr_file': {
+        assert(
+          certDetails.component === 'central-scan' ||
+            certDetails.component === 'scan',
+          'Signing machine cert for CVR file should be a VxCentralScan or VxScan cert'
+        );
+        break;
+      }
+      case 'election_definition': {
+        assert(
+          certDetails.component === 'admin',
+          'Signing machine cert for election definition should be a VxAdmin cert'
+        );
+        break;
+      }
+      /* istanbul ignore next: Compile-time check for completeness */
+      default: {
+        throwIllegalValue(artifact.type);
+      }
+    }
+    await verifyFirstCertWasSignedBySecondCert(
+      signingMachineCert,
+      this.vxCertAuthorityCertPath
+    );
+
+    const signingMachinePublicKey = await extractPublicKeyFromCert(
+      signingMachineCert
+    );
+    await verifySignature({
+      message,
+      messageSignature,
+      publicKey: signingMachinePublicKey,
+    });
+  }
+
+  private serializeSignatureFileContents(
+    signatureFileContents: SignatureFileContents
+  ): Buffer {
+    const { signature, signingMachineCert } = signatureFileContents;
+    return Buffer.concat([
+      // ECC signature length can vary ever so slightly, hence the need to persist length metadata
+      Buffer.from([signature.length]),
+      signature,
+      signingMachineCert,
+    ]);
+  }
+
+  private deserializeSignatureFileContents(
+    file: Buffer
+  ): SignatureFileContents {
+    const signatureLength = file[0];
+    assert(signatureLength !== undefined);
```

Comment:
let's also assert that the length is within some tight bounds we expect.

and while we're at it, that the total length of the file is at least long enough that we have a normal looking signingMachineCert. Paranoia paranoia here about potential parsing attacks :)

Reply:
For my understanding @benadida, what exactly would validation of the length of the `signingMachineCert` catch that wouldn't be caught during subsequent parsing of the cert? Is the thinking just that it's safer to short circuit early?

Reply:
To be clear this is just layers of defense and extra paranoia. I worry about to-be-discovered buffer bugs in OpenSSL that could be triggered by malicious inputs, and so any ways we can constrain what we feed OpenSSL is a good thing.

Reply:
Got it! That makes sense. Added both length checks accordingly: https://github.com/votingworks/vxsuite/pull/3476/commits/e32eb684370a543e9d1bf59c39de3fd27c502f84 https://github.com/votingworks/vxsuite/pull/3476/commits/5d4e48b111e21beb86fa541a94341c228dbaa6cc